File: app/src/main/java/com/amaury/pointage/v2/engine/SocialContributionCatalogV2.kt
package com.amaury.pointage.v2.engine

import kotlin.math.min

/**
 * Couche 1/6 - référentiel daté des retenues salariales légales de base.
 * Source 2026 : Urssaf, règles applicables au secteur privé.
 *
 * Important : chaque ligne conserve son taux ET son assiette. HoraTrack ne
 * remplace jamais la paie par un pourcentage global brut -> net.
 */
object SocialContributionCatalogV2 {
    enum class Base { GROSS, CSG_CRDS_2026, GROSS_CAPPED_MONTHLY_PASS }

    data class Rule(
        val id: String,
        val label: String,
        val employeeRate: Double,
        val base: Base,
        val validFromYear: Int,
        val validToYear: Int? = null,
        val source: String,
        val employerRate: Double = 0.0
    )

    data class Line(
        val id: String,
        val label: String,
        val baseAmount: Double,
        val rate: Double,
        val employeeAmount: Double,
        val source: String,
        val employerRate: Double = 0.0,
        val employerAmount: Double = 0.0
    )

    data class Estimate(
        val gross: Double,
        val employeeDeductions: Double,
        val netBeforeIncomeTax: Double,
        val lines: List<Line>,
        val warnings: List<String>,
        val employerContributions: Double = 0.0
    )

    private val rules2026 = listOf(
        Rule("old_age_uncapped", "Assurance vieillesse déplafonnée", 0.0040, Base.GROSS, 2026, 2026, "Urssaf - taux secteur privé 2026", employerRate = 0.0211),
        Rule("old_age_capped", "Assurance vieillesse plafonnée", 0.0690, Base.GROSS_CAPPED_MONTHLY_PASS, 2026, 2026, "Urssaf - taux secteur privé 2026", employerRate = 0.0855),
        Rule("csa_employer", "Contribution solidarité autonomie", 0.0, Base.GROSS, 2026, 2026, "Urssaf - taux secteur privé 2026", employerRate = 0.0030),
        Rule("social_dialogue_employer", "Contribution au dialogue social", 0.0, Base.GROSS, 2026, 2026, "Urssaf - taux secteur privé 2026", employerRate = 0.00016),
        Rule("csg_deductible", "CSG déductible", 0.0680, Base.CSG_CRDS_2026, 2026, 2026, "Urssaf - CSG/CRDS revenus d'activité 2026"),
        Rule("csg_taxable", "CSG imposable", 0.0240, Base.CSG_CRDS_2026, 2026, 2026, "Urssaf - CSG/CRDS revenus d'activité 2026"),
        Rule("crds", "CRDS", 0.0050, Base.CSG_CRDS_2026, 2026, 2026, "Urssaf - CSG/CRDS revenus d'activité 2026")
    )

    private val alsaceMoselle2026 = Rule(
        "alsace_moselle_local_health",
        "Régime local Alsace-Moselle — cotisation maladie supplémentaire",
        0.0130,
        Base.GROSS,
        2026,
        2026,
        "Urssaf / Régime Local Alsace-Moselle - taux 2026"
    )

    fun employeeRules(year: Int): List<Rule> = when (year) {
        2026 -> rules2026
        else -> emptyList()
    }

    /**
     * Assiette CSG/CRDS 2026 : le salaire bénéficie de l'abattement de 1,75 %
     * jusqu'à quatre fois le plafond social applicable, tandis que la part employeur
     * de protection sociale complémentaire soumise à CSG/CRDS est ajoutée ensuite,
     * sans appliquer cet abattement à cette part patronale.
     */
    private fun csgCrdsBase2026(
        gross: Double,
        applicableMonthlyPass: Double,
        employerProtectionCsgCrdsBaseAmount: Double
    ): Double {
        val cap = (applicableMonthlyPass * 4.0).coerceAtLeast(0.0)
        val abatedPart = min(gross, cap)
        val excess = (gross - cap).coerceAtLeast(0.0)
        return abatedPart * 0.9825 + excess + employerProtectionCsgCrdsBaseAmount
    }

    fun estimateEmployeeDeductions(
        gross: Double,
        year: Int,
        ceiling: SocialSecurityCeilingV2.Snapshot? = null,
        alsaceMoselleLocalRegime: Boolean? = null,
        employerProtectionCsgCrdsBaseAmount: Double? = null
    ): Estimate {
        val safeGross = gross.coerceAtLeast(0.0)
        val rules = employeeRules(year)
        if (rules.isEmpty()) {
            return Estimate(
                safeGross,
                0.0,
                safeGross,
                emptyList(),
                listOf("Cotisations salariales : barème non intégré pour $year")
            )
        }

        val validEmployerProtectionCsgCrdsBase = employerProtectionCsgCrdsBaseAmount
            ?.takeIf { it.isFinite() && it >= 0.0 }
        val monthlyPass = ceiling?.applicableMonthly
            ?: SocialSecurityCeilingV2.fullMonthly(year)
            ?: Double.POSITIVE_INFINITY
        val baseLines = rules.map { rule ->
            val base = when (rule.base) {
                Base.GROSS -> safeGross
                Base.CSG_CRDS_2026 -> csgCrdsBase2026(
                    safeGross,
                    monthlyPass,
                    validEmployerProtectionCsgCrdsBase ?: 0.0
                )
                Base.GROSS_CAPPED_MONTHLY_PASS -> min(safeGross, monthlyPass)
            }
            Line(
                id = rule.id,
                label = rule.label,
                baseAmount = base,
                rate = rule.employeeRate,
                employeeAmount = base * rule.employeeRate,
                source = rule.source,
                employerRate = rule.employerRate,
                employerAmount = base * rule.employerRate
            )
        }
        val localLines = if (year == 2026 && alsaceMoselleLocalRegime == true && safeGross > 0.0) {
            listOf(
                Line(
                    alsaceMoselle2026.id,
                    alsaceMoselle2026.label,
                    safeGross,
                    alsaceMoselle2026.employeeRate,
                    safeGross * alsaceMoselle2026.employeeRate,
                    alsaceMoselle2026.source
                )
            )
        } else emptyList()
        val lines = baseLines + localLines
        val total = lines.sumOf { it.employeeAmount }
        return Estimate(
            gross = safeGross,
            employeeDeductions = total,
            netBeforeIncomeTax = (safeGross - total).coerceAtLeast(0.0),
            lines = lines,
            warnings = buildList {
                add("Couche 1/6 : ce net est volontairement partiel.")
                add("Les parts patronales vieillesse, CSA et dialogue social 2026 sont intégrées séparément ; les autres cotisations patronales légales de base restent à compléter.")
                when {
                    employerProtectionCsgCrdsBaseAmount == null -> add("Assiette CSG/CRDS : part employeur de protection sociale complémentaire à confirmer, même si elle est nulle ; le sous-total courant reste calculé sur le brut connu uniquement.")
                    validEmployerProtectionCsgCrdsBase == null -> add("Assiette CSG/CRDS : part employeur de protection sociale complémentaire invalide ; aucune valeur n'est inventée.")
                    else -> add("Assiette CSG/CRDS : part employeur de protection sociale complémentaire confirmée ajoutée après l'abattement applicable au salaire.")
                }
                add("Retraite complémentaire, CEG/CET, mutuelle/prévoyance, convention et retenues propres à l'entreprise sont traitées dans les couches suivantes.")
                if (year == 2026 && alsaceMoselleLocalRegime == null) {
                    add("Régime local Alsace-Moselle : affiliation à confirmer ; aucune cotisation locale n'est inventée.")
                }
                if (year == 2026 && alsaceMoselleLocalRegime == true) {
                    add("Régime local Alsace-Moselle confirmé : cotisation salariale maladie supplémentaire de 1,30 % appliquée au brut déplafonné.")
                }
                ceiling?.warnings?.let(::addAll)
            }.distinct(),
            employerContributions = lines.sumOf { it.employerAmount }
        )
    }
}

File: ios/HPTravail/HPTravail/SalaryV2/SocialContributionCatalogV2.swift
import Foundation

/// Couche légale nationale iOS des retenues salariales de base.
///
/// Les taux sont datés. HoraTrack ne remplace jamais la paie par un pourcentage
/// global brut -> net et ne transforme jamais une donnée inconnue en zéro confirmé.
enum SocialContributionCatalogV2 {
    enum Base: Equatable {
        case gross
        case csgCrds2026
        case grossCappedMonthlyPass
    }

    struct Rule: Equatable {
        let id: String
        let label: String
        let employeeRate: Double
        let base: Base
        let validFromYear: Int
        let validToYear: Int?
        let source: String
        let employerRate: Double
    }

    struct Line: Equatable {
        let id: String
        let label: String
        let baseAmount: Double
        let rate: Double
        let employeeAmount: Double
        let source: String
        let employerRate: Double
        let employerAmount: Double
    }

    struct Estimate: Equatable {
        let gross: Double
        let employeeDeductions: Double
        let netBeforeIncomeTax: Double
        let lines: [Line]
        let warnings: [String]
        let employerContributions: Double
    }

    private static let rules2026: [Rule] = [
        Rule(
            id: "old_age_uncapped",
            label: "Assurance vieillesse déplafonnée",
            employeeRate: 0.0040,
            base: .gross,
            validFromYear: 2026,
            validToYear: 2026,
            source: "Urssaf — taux secteur privé 2026",
            employerRate: 0.0211
        ),
        Rule(
            id: "old_age_capped",
            label: "Assurance vieillesse plafonnée",
            employeeRate: 0.0690,
            base: .grossCappedMonthlyPass,
            validFromYear: 2026,
            validToYear: 2026,
            source: "Urssaf — taux secteur privé 2026",
            employerRate: 0.0855
        ),
        Rule(
            id: "csa_employer",
            label: "Contribution solidarité autonomie",
            employeeRate: 0,
            base: .gross,
            validFromYear: 2026,
            validToYear: 2026,
            source: "Urssaf — taux secteur privé 2026",
            employerRate: 0.0030
        ),
        Rule(
            id: "social_dialogue_employer",
            label: "Contribution au dialogue social",
            employeeRate: 0,
            base: .gross,
            validFromYear: 2026,
            validToYear: 2026,
            source: "Urssaf — taux secteur privé 2026",
            employerRate: 0.00016
        ),
        Rule(
            id: "csg_deductible",
            label: "CSG déductible",
            employeeRate: 0.0680,
            base: .csgCrds2026,
            validFromYear: 2026,
            validToYear: 2026,
            source: "Urssaf — CSG/CRDS revenus d'activité 2026",
            employerRate: 0
        ),
        Rule(
            id: "csg_taxable",
            label: "CSG imposable",
            employeeRate: 0.0240,
            base: .csgCrds2026,
            validFromYear: 2026,
            validToYear: 2026,
            source: "Urssaf — CSG/CRDS revenus d'activité 2026",
            employerRate: 0
        ),
        Rule(
            id: "crds",
            label: "CRDS",
            employeeRate: 0.0050,
            base: .csgCrds2026,
            validFromYear: 2026,
            validToYear: 2026,
            source: "Urssaf — CSG/CRDS revenus d'activité 2026",
            employerRate: 0
        )
    ]

    private static let alsaceMoselle2026 = Rule(
        id: "alsace_moselle_local_health",
        label: "Régime local Alsace-Moselle — cotisation maladie supplémentaire",
        employeeRate: 0.0130,
        base: .gross,
        validFromYear: 2026,
        validToYear: 2026,
        source: "Urssaf / Régime Local Alsace-Moselle — taux 2026",
        employerRate: 0
    )

    static func employeeRules(year: Int) -> [Rule] {
        year == 2026 ? rules2026 : []
    }

    static func estimateEmployeeDeductions(
        gross: Double,
        year: Int,
        ceiling: SocialSecurityCeilingV2.Snapshot? = nil,
        alsaceMoselleLocalRegime: Bool? = nil,
        employerProtectionCsgCrdsBaseAmount: Double? = nil
    ) -> Estimate {
        let safeGross = gross.isFinite ? max(0, gross) : 0
        let rules = employeeRules(year: year)
        guard !rules.isEmpty else {
            return Estimate(
                gross: safeGross,
                employeeDeductions: 0,
                netBeforeIncomeTax: safeGross,
                lines: [],
                warnings: ["Cotisations salariales : barème non intégré pour \(year)"],
                employerContributions: 0
            )
        }

        let validEmployerProtection: Double?
        if let amount = employerProtectionCsgCrdsBaseAmount, amount.isFinite, amount >= 0 {
            validEmployerProtection = amount
        } else {
            validEmployerProtection = nil
        }

        let monthlyPass = ceiling?.applicableMonthly
            ?? SocialSecurityCeilingV2.fullMonthly(year: year)
            ?? Double.infinity

        var lines = rules.map { rule -> Line in
            let baseAmount: Double
            switch rule.base {
            case .gross:
                baseAmount = safeGross
            case .csgCrds2026:
                baseAmount = csgCrdsBase2026(
                    gross: safeGross,
                    applicableMonthlyPass: monthlyPass,
                    employerProtectionCsgCrdsBaseAmount: validEmployerProtection ?? 0
                )
            case .grossCappedMonthlyPass:
                baseAmount = min(safeGross, monthlyPass)
            }

            return Line(
                id: rule.id,
                label: rule.label,
                baseAmount: baseAmount,
                rate: rule.employeeRate,
                employeeAmount: baseAmount * rule.employeeRate,
                source: rule.source,
                employerRate: rule.employerRate,
                employerAmount: baseAmount * rule.employerRate
            )
        }

        if year == 2026, alsaceMoselleLocalRegime == true, safeGross > 0 {
            lines.append(
                Line(
                    id: alsaceMoselle2026.id,
                    label: alsaceMoselle2026.label,
                    baseAmount: safeGross,
                    rate: alsaceMoselle2026.employeeRate,
                    employeeAmount: safeGross * alsaceMoselle2026.employeeRate,
                    source: alsaceMoselle2026.source,
                    employerRate: 0,
                    employerAmount: 0
                )
            )
        }

        let employeeTotal = lines.reduce(0) { $0 + $1.employeeAmount }
        let employerTotal = lines.reduce(0) { $0 + $1.employerAmount }
        var warnings = [
            "Couche légale iOS : ce net est volontairement partiel tant que les couches complémentaires ne sont pas raccordées.",
            "Les parts patronales vieillesse, CSA et dialogue social 2026 sont exposées séparément ; les autres cotisations patronales légales restent hors de ce sous-total."
        ]

        switch employerProtectionCsgCrdsBaseAmount {
        case nil:
            warnings.append(
                "Assiette CSG/CRDS : part employeur de protection sociale complémentaire à confirmer, même si elle est nulle ; le sous-total courant reste calculé sur le brut connu uniquement."
            )
        case let amount? where !amount.isFinite || amount < 0:
            warnings.append(
                "Assiette CSG/CRDS : part employeur de protection sociale complémentaire invalide ; aucune valeur n'est inventée."
            )
        default:
            warnings.append(
                "Assiette CSG/CRDS : part employeur de protection sociale complémentaire confirmée ajoutée après l'abattement applicable au salaire."
            )
        }

        warnings.append(
            "Retraite complémentaire, CEG/CET, mutuelle/prévoyance, convention et retenues propres à l'entreprise sont traitées dans les couches suivantes."
        )

        if year == 2026, alsaceMoselleLocalRegime == nil {
            warnings.append(
                "Régime local Alsace-Moselle : affiliation à confirmer ; aucune cotisation locale n'est inventée."
            )
        }
        if year == 2026, alsaceMoselleLocalRegime == true {
            warnings.append(
                "Régime local Alsace-Moselle confirmé : cotisation salariale maladie supplémentaire de 1,30 % appliquée au brut déplafonné."
            )
        }
        if let ceiling {
            warnings.append(contentsOf: ceiling.warnings)
        }

        return Estimate(
            gross: safeGross,
            employeeDeductions: employeeTotal,
            netBeforeIncomeTax: max(0, safeGross - employeeTotal),
            lines: lines,
            warnings: unique(warnings),
            employerContributions: employerTotal
        )
    }

    private static func csgCrdsBase2026(
        gross: Double,
        applicableMonthlyPass: Double,
        employerProtectionCsgCrdsBaseAmount: Double
    ) -> Double {
        let cap = max(0, applicableMonthlyPass * 4)
        let abatedPart = min(gross, cap)
        let excess = max(0, gross - cap)
        return abatedPart * 0.9825 + excess + employerProtectionCsgCrdsBaseAmount
    }

    private static func unique(_ values: [String]) -> [String] {
        var seen = Set<String>()
        return values.filter { seen.insert($0).inserted }
    }
}
